WordPress/gutenberg
# Split control for URL and Text within Link UI

## Description

<img width="593" alt="Screen Shot 2021-08-03 at 16 34 55" src="https://user-images.githubusercontent.com/444434/128044046-26d20b47-7912-47fd-881c-a4a790d3eb3f.png">

This PR adds a text field to the Link UI. When **editing an _existing_ link** are now two separate `<input>`s within the Link UI provided by the `<LinkControl>` component for both:

* link URL
* **link _text_** (new!)

### ⚠️ Please note 

- the text input is not displayed upon initial link creation. Rather it only appears when you wish to modify an existing link.
- formatting is **not preserved** when modifying the link text. This is expected and the rationale for it is explained by @jasmussen in [this comment](https://github.com/WordPress/gutenberg/pull/35457#issuecomment-940806880).
- yes the PR is large but this is mostly tests. I have offered to split it up but folks feel it's probably ok to keep everything in context.

This change effects:

* Any usages of the `core/link` format type - that's pretty much all usage of `RichText`.
* The Navigation block.

Whilst this UX slightly deviates from the "direct manipulation" paradigm, I believe that in this case it is both necessary and proportionate due to the unique UX considerations around the creation and modification of links.

To quote from [the Issue]( https://github.com/WordPress/gutenberg/issues/30170.):

> When you insert a link block such as a Menu Item or a Button (to a lesser extent, a Social Link), for the block to be functional it needs to have both a URL and a label configured. At the moment, this happens in a single link dialog interface.

> If you type arbitrary text, you are meant to be searching for a page or post that you can choose from the dropdown, or create a draft with that title. For example if you type "WordPress", if you just press Enter, you will create a menu item labelled "WordPress" which links to WordPress.

This is based on my original proposal in https://github.com/WordPress/gutenberg/pull/19413 which never got anywhere.

Fixes https://github.com/WordPress/gutenberg/issues/30170
Closes https://github.com/WordPress/gutenberg/issues/22435




## How has this been tested?

### Manual testing

* Create a new Post.
* Add some text.
* Add a link to a selection of that text - you should _only_ see the `URL` field when first creating the link. No text field.
* Click away and then click back on the link you just created. Click `Edit` in the Link UI.
* You should see x2 fields:
    - Text
    - URL
* Try changing the text and submitting (either by click or hitting `ENTER`). You should see the underlying link text in the paragraph block change to match the contents of the Link UI's text field.
* Repeat but try:
    - including spaces in your link
    - creating the link at the beginning/end of a piece of text.
    - using the keyboard to select the link

### Automated tests

Run the unit tests:

```
npm run test-unit packages/block-editor/src/components/link-control/
```

and also:

```
npm run test-unit packages/format-library/src/link/test
```

Run the e2e tests:

```
npm run test-e2e:watch packages/e2e-tests/specs/editor/various/links.test.js
```



## Screenshots 

https://user-images.githubusercontent.com/444434/137137231-9282756b-ac05-4240-9c5c-ac07ce1e0d3c.mov

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
